harness/determined/common/api/authentication.py

```diff
@@ -62,139 +69,79 @@ def default_load_user_password(
-    password: str,
+def login_with_cache(
+    master_address: Optional[str] = None,
+    requested_user: Optional[str] = None,
+    password: Optional[str] = None,
     cert: Optional[certs.Cert] = None,
-) -> str:
-    password = api.salt_and_hash(password)
-    unauth_session = api.Session(user=username, master=master_address, auth=None, cert=cert)
-    login = bindings.v1LoginRequest(username=username, password=password, isHashed=True)
-    r = bindings.post_Login(session=unauth_session, body=login)
-    token = r.token
-
-    return token
-
-
-class LogoutAuthentication(Authentication):
+) -> UsernameTokenPair:
     """
-    An api-compatible Authentication object that is basically exactly a UserTokenPair.
+    Log in, preferring cached credentials in the TokenStore, if possible.
 
-    TODO(MLG-215): delete Authentication class and write a function that returns a UsernameTokenPair
-    in its place, and let do_request() take UsernameTokenPair as input.
+    This is the login path for nearly all user-facing cases.
     """
 
-    def __init__(self, session_user: str, session_token: str) -> None:
-        self.session_user = session_user
-        self.session_token = session_token
+    master_address = master_address or util.get_default_master_address()
+    token_store = TokenStore(master_address)
 
-    def get_session_user(self) -> str:
-        return self.session_user
+    # Get session_user and password given the following priority:
+    # 1. User passed in with flag (requested_user)
+    # 2. User from environment if DET_PASS is set.
+    # 3. Active user from the token store.
+    user, password = default_load_user_password(requested_user, password, token_store)
+
+    # For login, we allow falling back to the default username.
+    if not user:
+        user = constants.DEFAULT_DETERMINED_USER
+    assert user is not None
+
+    # Check the token store if this session_user has a cached token. If so, check with the
+    # master to verify it has not expired. Otherwise, let the token be None.
+    token = token_store.get_token(user)
+    if token is not None and not _is_token_valid(master_address, token, cert):
+        token_store.drop_user(user)
+        token = None
+
+    if token is not None:
+        return UsernameTokenPair(user, token)
+
+    # Special case: use token provided from the container environment if:
+    # - No token was obtained from the token store already,
+    # - There is a token available from the container environment, and
+    # - No user was explicitly requested, or the requested user matches the token available in the
+    #   container environment.
+    if (
+        util.get_det_username_from_env() is not None
+        and util.get_det_user_token_from_env() is not None
+        and requested_user in (None, util.get_det_username_from_env())
+    ):
+        user = util.get_det_username_from_env()
+        assert user
+        token = util.get_det_user_token_from_env()
+        assert token
```

Comment:
No, this mechanism is an internal mechanism by which the master configures the containers it launches to automatically begin in a logged in state.  It is not validating a user input.

Reply:
I believe your argument is that the `DET_USER_TOKEN` is not a public interface, and therefore we should treat it like a special-cased backdoor from master that has very specific semantics.

Reply:
yes, exactly

Reply:
I get it, but it feels to me like we made a mistake somewhere implementing that special-case authentication this way. In the event you feel the same (or I can convince you of it), I can't quite say whether a small refactor should fall into this PR, go in a ticket, or get a handwavy comment as a part of this PR.

I've asked about a thousand times, sometimes rhetorically sometimes not, but what is "logging in"? I _think_ we've decided that "logging in" is getting a token that can can sign later calls (there's also stuff that might happen in the DB, but I think we're taking the view that that stuff is irrelevant and not worth thinking about here in the frontend).

Sometimes logging in means getting a token from the token store, sometimes from master, and here from the environment. Sometimes that token in stored in a token store; sometimes it isn't.

Since the `DET_USER_TOKEN` case isn't public and we don't need to support it, it'd be great if that special case eliminated. Could we get the `DET_USER_TOKEN` into the token store for `DET_USER`? Then I think that logic is already handled with existing defaluts.

It also seems a little weird to me for the `login_with_cache` semantics to be variable wrt the token store. In other paths we set tokens after receiving them.

Reply:
Oh. I see I've already said as much. https://github.com/determined-ai/determined/pull/8347#discussion_r1482265449

At least I'm consistent. I'll ping you for a call.

Reply:
So, it sounds like what you are suggesting could be pre-packing an auth.json from the master into the task container.

Technically, this could replace the environment variables strategy, but it is hard or sometimes impossible for us to predict where $HOME will be inside the container before the container runs, such as if somebody runs a container with an LDAP socket bound into the container.

The environment variable strategy is trivial and supported by all container runtimes, so it's a very safe bet.

And anything other than pre-packing an auth.json (like using some well-known filename instead of the environment variable) would have basically identical semantics to the environment variable name, so I don't see a lot of value in non-environment, non-auth.json alternatives either.

Reply:
Thanks for all the conversation. Your explanation and claim of "standard practice" has helped.

Also helped, writing out the fallback logic:

1. Infer the user
2. If we have a token for that inferred user, return it
3. If the inferred user equals an environment variable and the token is also in the environment, return it
4. If the inferred user is a special string and no password is set, infer the password (perhaps by asking the user)
5. log in with the inferred user and inferred password

The code is complicated, but the logic isn't quite as bad as I'd remembered it.

I think this gets better if we move "3" to above "1". If you agree, I'll add it to the wiki.

Reply:
It doesn't get better, it introduces a bug if you move 3 above 1.

It is a bug because users who run `det user login` inside the container followed by `det user whoami`, they will always see the injected container user instead of the user they just logged in as.

Fun fact, that's a real bug we had to fix at some point.

Reply:
Working this out for myself, post-PR-8347 this goes:

1. `det user whoami`
   1. `cli.setup_session`
      1. `authentication.login_with_cache(requested_user=None)`
         1. `authentication.default_load_user_password`
            1. (`DET_USER`, `DET_PASS`) if these both exist 
            1. `TokenStore.active_user` otherwise
         2. use `DET_USER` & `DET_USER_TOKEN` if we never got a user from above

ok! I get it!

Thanks for the insight.

This shows me there's a dependency on no `DET_PASS` being present to get the `DET_USER_TOKEN`, but this is kind of gross already, and I don't think we need to do anything about that.

Ok. I'm done with this part.